datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/BatchAddTagsResolver.java

```diff
@@ -52,6 +68,39 @@ public CompletableFuture<Boolean> get(DataFetchingEnvironment environment) throw
     });
   }
 
+  // If updating schema field tags fails, try again on a sibling until there are no more siblings to try. Then throw if necessary.
+  private Boolean attemptBatchAddTagsWithSiblings(
+      @Nonnull final List<Urn> tagUrns,
+      @Nonnull final ResourceRefInput resource,
+      @Nonnull final QueryContext context,
+      @Nonnull final HashSet<Urn> attemptedUrns,
+      @Nonnull final List<Urn> siblingUrns
+  ) {
+    attemptedUrns.add(UrnUtils.getUrn(resource.getResourceUrn()));
+    final List<ResourceRefInput> resources = new ArrayList<>();
+    resources.add(resource);
+
+    try {
+      validateInputResources(resources, context);
+      batchAddTags(tagUrns, resources, context);
+      return true;
+    } catch (Exception e) {
+      final Optional<Urn> siblingUrn = SiblingsUtils.getNextSiblingUrn(siblingUrns, attemptedUrns);
+
+      if (siblingUrn.isPresent()) {
+        log.warn(
+            "Failed to add tags for resourceUrn {} and subResource {}, trying sibling urn {} now.",
+            resource.getResourceUrn(), resource.getSubResource(), siblingUrn.get()
+        );
+        resource.setResourceUrn(siblingUrn.get().toString());
+        return attemptBatchAddTagsWithSiblings(tagUrns, resource, context, attemptedUrns, siblingUrns);
```

Comment:
Wow so this can become recursive. Very interesting. No chance this leads to an infinite loop, right?

Reply:
correct! yeah so I only query once to get the siblings of the given urn in input. Then I only loop over those a total of once and probably not even the whole list. I just keep trying until either one succeeds or if we have tried all the original siblings given (I keep track of the attemptedUrns in a set and make sure we don't repeat thanks to the helper function `getNextSiblingUrn`)